util/cpuprofile/pprof_api.go

```diff
@@ -0,0 +1,196 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cpuprofile
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/pprof/profile"
+	goutil "github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// ProfileHTTPHandler is same as pprof.Profile.
+// The difference is ProfileHTTPHandler uses cpuprofile.GetCPUProfile to fetch profile data.
+func ProfileHTTPHandler(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	sec, err := strconv.ParseInt(r.FormValue("seconds"), 10, 64)
+	if sec <= 0 || err != nil {
+		sec = 30
+	}
+
+	if durationExceedsWriteTimeout(r, float64(sec)) {
+		serveError(w, http.StatusBadRequest, "profile duration exceeds server's WriteTimeout")
+		return
+	}
+
+	// Set Content Type assuming StartCPUProfile will work,
+	// because if it does it starts writing.
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	pc := NewCollector()
+	err = pc.StartCPUProfile(w)
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+		return
+	}
+	time.Sleep(time.Second * time.Duration(sec))
+	err = pc.StopCPUProfile()
+	if err != nil {
+		serveError(w, http.StatusInternalServerError, "Could not enable CPU profiling: "+err.Error())
+	}
+}
+
+const (
+	labelSQLDigest  = "sql_digest"
+	labelPlanDigest = "plan_digest"
+)
+
+// Collector is a cpu profile collector, it collect cpu profile data from globalCPUProfiler.
+type Collector struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	started   atomic.Bool
+	firstRead chan struct{}
+	wg        sync.WaitGroup
+
+	dataCh   ProfileConsumer
+	profiles []*ProfileData
+	writer   io.Writer
+}
+
+// NewCollector returns a new NewCollector.
+func NewCollector() *Collector {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Collector{
+		ctx:       ctx,
+		cancel:    cancel,
+		firstRead: make(chan struct{}),
+		dataCh:    make(ProfileConsumer, 1),
+	}
+}
+
+// StartCPUProfile is a substitute for the `pprof.StartCPUProfile` function.
+// You should use this function instead of `pprof.StartCPUProfile`.
+// Otherwise you may fail, or affect the TopSQL feature and pprof profile HTTP API .
+func (pc *Collector) StartCPUProfile(w io.Writer) error {
+	if !pc.started.CAS(false, true) {
```

Comment:
How about simply adding a lock or does not handle concurrent at all? I see that your `StopCPUProfile` is not thread-safe.

Reply:
done.